### PR TITLE
Clarify return type in `training_step` docs in case of manual optimization

### LINKED ISSUE #19311: Enable to return outputs from {training,validation,test}_step in Manual Optimization 🙏
### Description & Motivation

In previous Lightning versions, it was possible to use multiple optimizers with automatic optimization. Since lightning>=2.0.0 this is no longer possible (_i.e._ the argument `optimizer_idx` of the training step disappeared). The problem is not that we have to do the optimization manually, but that we cannot return anything from the {training,validation,test}_step method. Indeed, I have been using the dict return type a lot to put the logging logic inside Lightning hooks like `on_train_batch_end` which takes the outputs of `training_step` as input.

### Pitch

I would like to be able to return a dict without the key 'loss' from {training,validation,test}_step when using manual optimization.

### Alternatives

The only alternative I can find is to move my logging code that is written in `on_train_batch_end`, `on_validation_batch_end`, `on_test_batch_end` of an AbstractLightningModule into the GANLightningModule. The issue is that it generates duplicate code because I have other LightningModules that also inherit from AbstractLightningModule and use the same logging hooks.

### Additional context

PS: Is there a typo in [this line](https://lightning.ai/docs/pytorch/2.1.2/common/lightning_module.html#training-step:~:text=only%20supported%20for-,automatic,-optimization.) of the documentation ? Shouldn't it be _" Skip to the next batch. This is only supported for ~automatic~ **manual** optimization."_

cc @borda

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes a small confusion in #19311, by clarifying what returning None means. 
Closes #19311

----
📚 Documentation preview 📚: https://pytorch-lightning--19327.org.readthedocs.build/en/19327/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -690,8 +690,8 @@ def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
         Return:
             - :class:`~torch.Tensor` - The loss tensor
             - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
```
Comment (jhauret) [inline src/lightning/pytorch/core/module.py:692]:
```suggestion
            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'`` in the case of automatic optimization.
```

**Review by carmocca — APPROVED** [review]
I would keep the comment about distributed not being supported but either way LGTM

File: src/lightning/pytorch/core/module.py [hunk 2]
```diff
@@ -689,9 +689,11 @@ def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
 
         Return:
             - :class:`~torch.Tensor` - The loss tensor
-            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
-            - ``None`` - Skip to the next batch. This is only supported for automatic optimization.
-                This is not supported for multi-GPU, TPU, IPU, or DeepSpeed.
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'`` in the case of
```
Comment (Borda) [inline src/lightning/pytorch/core/module.py:692]:
```suggestion
            - ``dict`` - A dictionary which can include any keys, but must include the key ``'loss'`` in the case of
```

**Review by Borda — APPROVED** [review]
